c2rust-analyze/src/context.rs

```diff
@@ -41,6 +41,70 @@ bitflags! {
         const OFFSET_SUB = 0x0020;
         /// This pointer can be freed.
         const FREE = 0x0040;
+
+        /// This pointer is non-null.
+        ///
+        /// [`NON_NULL`] is set (or not) when the pointer is created,
+        /// and it flows forward along dataflow edges.
```

Comment:
should it also flow backward? I'm thinking of a call to `offset`, e.g.

```
_1 = malloc(...);
_2 = std::ptr::offset(_1, ...);
something_else_with(_1); // pointer permissions for NON_NULL may be useful here
```

`_2` would get the `NON_NULL` permission and so would `_1`. It seems like this should happen regardless of the checks the initial C code made

Reply:
I guess we had a race condition :)

In your example, it's fine to propagate backward, but if you were to wrap everything after the `malloc` in an `if !_1.is_null()`, then it wouldn't make sense to propagate `NON_NULL` outside that `if`.  And we don't currently have a notion of "this pointer X has permission Y, but only inside this block" - the pointer either has the permission or it doesn't.

Reply:
based on other considerations with relaxing permissions, it sounds like it should not propagate backwards unless it's very clear there is no control flow, which we don't analyze anyway